First commit of Lucas Number generator.

andrewvanderbyl opened the pull request:
This project generates Lucas numbers. The sequence is limited by the
maximum Lucas number required (threshold reached). The golden ratio is
also calculated for good measure however not tested for in the unit
test as a correct Lucas sequence by default computes to the correct
ratio. The unit test only looks to the first 10 enties for validation.

andrewvanderbyl committed:
First commit of Lucas Number generator.

This project generates Lucas numbers. The sequence is limited by the
maximum Lucas number required (threshold reached). The golden ratio is
also calculated for good measure however not tested for in the unit
test as a correct Lucas sequence by default computes to the correct
ratio. The unit test only looks to the first 10 enties for validation.

james-smith-za commented on .vscode/launch.json:
```diff
@@ -8,7 +8,7 @@
             "type": "python",
             "request": "launch",
-            "program": "${workspaceFolder}/SigAverager/testing/test_dsp_signal_averager_decorator.py",
+            "program": "${workspaceFolder}/LucasNumber/testing/test_lucas_number.py",
```
Again, suggest adding this to your `.gitignore` file.

I'd also suggest caution against staging things for commit too eagerly. I typically review each file that I add.

andrewvanderbyl replied:
I usually include this so that if testing in VS code, the launch.json file specific to the project is there and setup for use.

james-smith-za replied:
That's... an interesting approach. With personal repos that's quite a good idea, but I wouldn't suggest using it for organisation ones, unless we want to strictly mandate the use of a particular IDE / debugger, and I'm not sure we want to do that.

james-smith-za commented on LucasNumber/testing/test_lucas_number.py:
```diff
@@ -0,0 +1,54 @@
+    print("Done!")
+
+    # Iterate through the generated sequence and compare to the expected sequence.
+    for i in range(len(expected_lucas_sequence)):
+        if generated_sequence[i][0] == expected_lucas_sequence[i]:
+            test = True
+        else:
+            test = False
+            break
+    assert test
```
Generally you'd just `assert` your tested condition directly. This makes the output of the test more verbose - because you can see, when it fails, exactly where it fails.

```suggestion
    for i in range(len(expected_lucas_sequence)):
        assert generated_sequence[i][0] == expected_lucas_sequence[i]:
```

james-smith-za replied:
You can also compare lists directly:
```python
for lucas in lng:
    generated_sequence.append(lucas[0])  # You don't need the ratio in this sequence really.
    ratios.append(lucas[1])  # If you want to see them tending towards \phi, then you can plot this sequence.
...
assert generated_sequence == expected_lucas_sequence
```

The one advantage that your approach has, I'll grant, is that you'll see exactly which of the loops failed, but not when you are using the intermediate `test` variable. In this case, you just get something like this:
![image](https://user-images.githubusercontent.com/10864691/121846600-c67bc580-ccd6-11eb-9843-9df85c96a54a.png)

```

andrewvanderbyl replied:
Noted. Applied the assert change.

james-smith-za commented on LucasNumber/LucasNumber/lucas.py:
```diff
@@ -0,0 +1,44 @@
+    while True:
+        lucas_number = sequence[0] + sequence[1]
+        golden_ratio = golden_ratio_calc(sequence)
+        next_ln = yield [lucas_number, golden_ratio]
```
Nice usage of the co-routine idea.

You may just want to include in your function documentation that the user needs to send the next number back into the coroutine, and close when they are finished, because this isn't obvious from your function documentation.

The only reservation that I have, and this is more of a nit-pick (or requirements interpretation than anything else) is that in this case, generating the right sequence involves the user's code `send`ing the correct number back into the co-routine.

So there's nothing wrong with the code, in fact this may be exactly what is needed in some hypothetical situations, it's just not quite what I had in mind. This is more of a requirements analysis issue than anything else.

andrewvanderbyl replied:
Noted. The intention was to try use both features (yield, send) more than trying to produce a production piece of code.

james-smith-za replied:
Yes, and as such this worked well. Hence why I said more of a nit-pick. I wouldn't suggest making any changes here, you've got the general generator concept quite well.

james-smith-za commented on LucasNumber/LucasNumber/__init__.py:
```diff
@@ -0,0 +1,4 @@
+"""Import the following modules for the Lucas number generator package."""
+from .lucas import lucas_number_generator
+
+__all__ = ["lucas_number_generator", "golden_ratio_calc"]
```
You need to import `golden_ratio_calc` as well if you're going to make it available in `__all__`.

andrewvanderbyl replied:
Noted. Import added.

james-smith-za reviewed (changes requested):
Nice, overall! Just some small suggestions.

andrewvanderbyl committed:
Made changes as per PR with JSmith.

__init__.py : Added import for golden_ratio_calc
test_lucas_number.py : Chnaged assert for unit test when verifying
lucas sequence.

Added signal averager using decorators. These are separate files rather than an overwrite so that a comparison can be made easily. The intention is not to merge into one file but keep as separate examples.

Added fix to SigAverager using decorators

This commit added a fix to the recursion of the return value used when
doing signal averaging.

No SARAO JIRA ticket as it's a private repo.

Added a realisation to the README. It should only to this branch.
